### PR TITLE
Stress watcher incorporating scheduling rules for chaos-mesh 2.1.5

### PR DESCRIPTION
**Pull request by ckairen**
Also added in Unit testing for stress watcher resources

### CODE REVIEW

File: tools/stress-cluster/services/Stress.Watcher/src/ChaosResources/Resource.cs [hunk 1]
```diff
@@ -34,6 +35,45 @@ public class ChaosResourceSpec
...
+                } else if (property.PropertyType == typeof(ChaosResourceSpec)) {
+                    ChaosResourceSpec chaosSpec = (ChaosResourceSpec)property.GetValue(this, null);
+                    string ti = chaosSpec?.GetTestInstance();
+                    testInstance = string.IsNullOrEmpty(ti)?testInstance:ti;
+                }
+                if (!string.IsNullOrEmpty(testInstance)) {
+                    return testInstance;
+                }
+            }
+            return testInstance;
```
Comment (benbp) [inline tools/stress-cluster/services/Stress.Watcher/src/ChaosResources/Resource.cs:75]: Would it be better to throw here?

File: tools/stress-cluster/services/Stress.Watcher/src/ChaosResources/Resource.cs [hunk 2]
```diff
@@ -34,6 +35,45 @@ public class ChaosResourceSpec
...
+        public string GetTestInstance()
+        {
+            string testInstance = "";
+            foreach (PropertyInfo property in this.GetType().GetProperties()) {
+                if (property.PropertyType == typeof(ChaosSelector)) {
+                    testInstance = Selector?.LabelSelectors?.TestInstance;
+                } else if (property.PropertyType == typeof(ChaosResourceSpec)) {
+                    ChaosResourceSpec chaosSpec = (ChaosResourceSpec)property.GetValue(this, null);
+                    string ti = chaosSpec?.GetTestInstance();
+                    testInstance = string.IsNullOrEmpty(ti)?testInstance:ti;
```
Comment (benbp) [inline tools/stress-cluster/services/Stress.Watcher/src/ChaosResources/Resource.cs:69]: I think this no space ternary is a little confusing to read.

File: tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs [hunk 3]
```diff
@@ -164,10 +164,15 @@ public async Task StartChaosResources(V1Pod pod)
                         .Where(cr => ShouldStartChaos(cr, pod))
                         .Select(async cr =>
                         {
+                            string plural = "";
+                            foreach (string pluralName in Enum.GetNames(typeof(GenericChaosClient.ChaosResourcePlurals))) {
+                                if (pluralName.Contains(cr.Kind.ToLower())) {
+                                    plural = pluralName;
```
Comment (benbp) [inline tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs:170]:
```suggestion
                                    plural = pluralName;
                                    break;
```

File: tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs [hunk 4]
```diff
@@ -164,10 +164,15 @@ public async Task StartChaosResources(V1Pod pod)
...
+                                }
+                            }
                             await Client.PatchNamespacedCustomObjectWithHttpMessagesAsync(
                                     PodChaosResumePatchBody, ChaosClient.Group, ChaosClient.Version,
-                                    pod.Namespace(), cr.Kind.ToLower(), cr.Metadata.Name);
-
+                                    pod.Namespace(), plural, cr.Metadata.Name);
```
Comment (benbp) [inline tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs:175]: Is this right? We update an individual object here so will the plurals API route still work?

File: tools/stress-cluster/services/Stress.Watcher/src/ChaosResources/Resource.cs [hunk 5]
```diff
@@ -34,6 +35,45 @@ public class ChaosResourceSpec
...
+        [JsonProperty(PropertyName = "kernelChaos")]
+        public ChaosResourceSpec KernelChaos { get; set; }
+
+        [JsonProperty(PropertyName = "timeChaos")]
+        public ChaosResourceSpec TimeChaos { get; set; }
+
+        [JsonProperty(PropertyName = "jvmChaos")]
+        public ChaosResourceSpec JvmChaos { get; set; }
+
+        public string GetTestInstance()
```
Comment (benbp) [inline tools/stress-cluster/services/Stress.Watcher/src/ChaosResources/Resource.cs:60]: This is a super nit, but I wonder if it would be better just to use a nullable string response type here, to differentiate between "testInstance label exists and is empty" vs. "no testInstance label." It would also simplify your below code a little bit.

File: tools/stress-cluster/services/Stress.Watcher/tests/ResourceTest.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            var networkChaos = new ChaosResourceSpec();
+            schedule.Spec = new ChaosResourceSpec();
+            networkChaos.Selector = new ChaosSelector() {
+                LabelSelectors = new ChaosLabelSelectors() {
+                    TestInstance = "ResourceTestTestInstance"
+                }
+            };
+            schedule.Spec.NetworkChaos = networkChaos;
+            networkChaos.GetTestInstance().Should().Be("ResourceTestTestInstance");
+            schedule.Spec?.GetTestInstance().Should().Be("ResourceTestTestInstance");
```
Comment (benbp) [inline tools/stress-cluster/services/Stress.Watcher/tests/ResourceTest.cs:26]: It would be good to add tests for empty schedule (no inner chaos spec) and missing label as well (no TestInstance label).

Comment (benbp) [reply tools/stress-cluster/services/Stress.Watcher/tests/ResourceTest.cs:26]: also, do schedules support multiple inner chaos specs?

Comment (ckairen) [reply tools/stress-cluster/services/Stress.Watcher/tests/ResourceTest.cs:26]: Correct me if I'm wrong, since I'm not well versed in go. The [source code](https://github.com/chaos-mesh/chaos-mesh/blob/master/api/v1alpha1/schedule_types.go#L74) seems to suggest that there can only be one type of chaos being scheduled.

File: tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs (hunk 4 above)
Comment (ckairen) [reply tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs:175]: Yup, positive. I switched to plural because "schedule" doesn't work. it needs "schedules". Before chaos-mesh 2.1.5, it was fine because plural and singular for XXXChaos are the same.

**Review by benbp — APPROVED** [review]
